File: 3. Behavioral Design Patterns/Memento/Java/MementoExample.java
// Memento Pattern Example in Java
// Text editor that can save and restore document states

import java.util.*;

class TextMemento {
    private final String content;
    public TextMemento(String content) { this.content = content; }
    public String getContent() { return content; }
}

class TextEditor {
    private String content = "";
    public void write(String text) { content += text; }
    public void clear() { content = ""; }
    public String getContent() { return content; }
    public TextMemento save() { return new TextMemento(content); }
    public void restore(TextMemento memento) { content = memento.getContent(); }
}

class EditorHistory {
    private final Stack<TextMemento> history = new Stack<>();
    public void save(TextEditor editor) { history.push(editor.save()); }
    public void undo(TextEditor editor) { if (!history.isEmpty()) editor.restore(history.pop()); }
}

public class MementoExample {
    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        EditorHistory history = new EditorHistory();

        editor.write("Hello ");
        history.save(editor);
        System.out.println("Content: " + editor.getContent());

        editor.write("World!");
        history.save(editor);
        System.out.println("Content: " + editor.getContent());

        history.undo(editor);
        System.out.println("After undo: " + editor.getContent());

        history.undo(editor);
        System.out.println("After undo: " + editor.getContent());
    }
}

File: 3. Behavioral Design Patterns/Memento/Python/MementoExample.py
# Memento Pattern Example in Python
# Text editor that can save and restore document states

class TextMemento:
    def __init__(self, content):
        self._content = content
    
    def get_content(self):
        return self._content

class TextEditor:
    def __init__(self):
        self._content = ""
    
    def write(self, text):
        self._content += text
    
    def clear(self):
        self._content = ""
    
    def get_content(self):
        return self._content
    
    def save(self):
        return TextMemento(self._content)
    
    def restore(self, memento):
        self._content = memento.get_content()

class EditorHistory:
    def __init__(self):
        self._history = []
    
    def save(self, editor):
        self._history.append(editor.save())
    
    def undo(self, editor):
        if self._history:
            memento = self._history.pop()
            editor.restore(memento)

def main():
    editor = TextEditor()
    history = EditorHistory()
    
    editor.write("Hello ")
    history.save(editor)
    print(f"Content: {editor.get_content()}")
    
    editor.write("World!")
    history.save(editor)
    print(f"Content: {editor.get_content()}")
    
    history.undo(editor)
    print(f"After undo: {editor.get_content()}")
    
    history.undo(editor)
    print(f"After undo: {editor.get_content()}")

if __name__ == "__main__":
    main()
